merlin-framework-processor/src/main/java/gov/nasa/jpl/aerie/merlin/processor/generator/MissionModelGenerator.java

```diff
@@ -646,6 +658,100 @@ public JavaFile generateActivityTypes(final MissionModelRecord missionModel) {
+                    .build())
+            .addModifiers(Modifier.PUBLIC, Modifier.FINAL);
+
+    record ComponentMapperNamePair(String componentName, String mapperName) {}
+    for (final var record : recordTypes) {
+      if (record.getKind() != ElementKind.RECORD) {
+        messager().printError("@%s is only allowed on records".formatted(AutoValueMapper.class.getSimpleName()), record);
+      }
+
+      final var methodName = record.toString().replace(".", "_");
+      final var componentToMapperName = new ArrayList<ComponentMapperNamePair>();
+      final var necessaryMappers = new HashMap<TypeMirror, String>();
+      for (final var element : record.getEnclosedElements()) {
+        if (!(element instanceof RecordComponentElement el)) continue;
+        final var typeMirror = el.getAccessor().getReturnType();
+        final var elementName = element.toString();
+        final var valueMapperIdentifier = typeMirror.toString().replace(".", "_") + "ValueMapper";
+        componentToMapperName.add(new ComponentMapperNamePair(elementName, valueMapperIdentifier));
+        necessaryMappers.put(typeMirror, valueMapperIdentifier);
+      }
+
+      final var typeRule = new TypeRule(
+          new TypePattern.ClassPattern(ClassName.get(ValueMapper.class), List.of(TypePattern.from(record.asType()))),
+          Set.of(),
+          necessaryMappers
+              .keySet()
+              .stream()
+              .map(component -> (TypePattern) new TypePattern.ClassPattern(
+                  ClassName.get(ValueMapper.class),
+                  List.of(new TypePattern.ClassPattern((ClassName) ClassName.get(component).box(), List.of()))))
+              .toList(),
+          typeName,
+          methodName);
+      typeRules.add(typeRule);
+
+      final var methodBuilder = MethodSpec
+          .methodBuilder(methodName)
+          .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
+          .returns(ParameterizedTypeName.get(ClassName.get(ValueMapper.class), TypeName.get(record.asType())))
+          .addParameters(
+              necessaryMappers
+                  .entrySet()
+                  .stream()
+                  .map(mapperRequest -> ParameterSpec
+                      .builder(
+                          ParameterizedTypeName.get(
+                              ClassName.get(ValueMapper.class),
+                              ClassName.get(mapperRequest.getKey()).box()),
+                          mapperRequest.getValue(),
+                          Modifier.FINAL)
+                      .build())
+                  .toList());
+
+      final var iter = componentToMapperName.iterator();
+      final var codeBlockBuilder = CodeBlock.builder();
+      codeBlockBuilder.add("return new $T<>(\n", RecordValueMapper.class);
+      codeBlockBuilder.add("  $L.class,\n", record.toString());
+      codeBlockBuilder.add("  $T.of(\n", List.class);
+
+      while (iter.hasNext()) {
```

Comment:
I'd replace this with:
```java
for (final var iter = componentToMapperName.iterator(); iter.hasNext();)
```

This whole block is cool with me but you could also do some variety of a `componentToMapperName.stream().map(...).reduce(...)` to inject the `","` text between each generated block. I think keeping this imperative actually makes the most sense for readability though.

Reply:
Hmm, I'm not convinced that using an iterator for loop is more readable than a while loop (I think it kind of obfuscates rather than clarifies what's going on)... however! By poking at this you've given me the motivation to go read the javapoet docs a little bit, and it turns out there's a `CodeBlock.join` method, which makes your second suggestion the best option, I think. That loop has been replaced with:

```java
codeBlockBuilder.add(
    CodeBlock.join(componentToMapperName
      .stream()
      .map(recordComponent ->
               CodeBlock
                   .builder()
                   .add("    new $T<>(\n", RecordValueMapper.Component.class)
                   .add("      $S,\n", recordComponent.componentName())
                   .add("      $T::$L,\n", ClassName.get(record), recordComponent.componentName())
                   .add("    $L)", recordComponent.mapperName())
                  .build())
      .toList(), ",\n"));
```

Reply:
Ohhh awesome! Thanks, this looks very clean.

Reply:
I had no idea `CodeBlock.join` existed. *Nice.*